core/sequencer/server.go

```diff
@@ -222,72 +223,58 @@ func (s *Server) unappliedMetric(ctx context.Context, directoryID string, maxCou
 func (s *Server) RunBatch(ctx context.Context, in *spb.RunBatchRequest) (*empty.Empty, error) {
-	defResp, err := s.loopback.DefineRevisions(ctx, &spb.DefineRevisionsRequest{
-		DirectoryId: in.DirectoryId,
-		MinBatch:    in.MinBatch,
-		MaxBatch:    in.MaxBatch})
+	revs, err := s.loopback.GetNewRevisions(ctx,
+		&spb.GetNewRevisionsRequest{DirectoryId: in.DirectoryId})
 	if err != nil {
 		return nil, err
 	}
 
-	unappliedRevisions.Set(float64(len(defResp.OutstandingRevisions)))
+	count, end := revs.End-revs.Begin, revs.End
+	unappliedRevisions.Set(float64(count))
+	// TODO(pavelkalinnikov): Don't share LogPublishBatchSize with Publish.
+	if mx := int64(s.LogPublishBatchSize); count > mx {
+		// Only handle up to LogPublishBatchSize revisions in one go.
+		glog.Warningf("%v: RunBatch: too many outstanding revisions: %d", in.DirectoryId, count)
+		end = revs.Begin + mx
+	}
 
-	var handledCount uint64
-	for _, rev := range defResp.OutstandingRevisions {
-		if handledCount == s.LogPublishBatchSize {
-			// Only handle up to LogPublishBatchSize revisions per batch.
-			glog.Errorf("RunBatch - Too many outstanding revisions to apply: %d", len(defResp.OutstandingRevisions))
-			break
-		}
+	for rev := revs.Begin; rev < end; rev++ {
 		revReq := &spb.ApplyRevisionRequest{DirectoryId: in.DirectoryId, Revision: rev}
-		_, err := s.loopback.ApplyRevision(ctx, revReq)
-		if err != nil {
+		if _, err := s.loopback.ApplyRevision(ctx, revReq); err != nil {
 			return nil, err
 		}
-		handledCount++
 	}
 	return &empty.Empty{}, nil
 }
 
-// DefineRevisions examines the outstanding mutations and returns a list of
-// outstanding revisions that have not been applied.
+// DefineRevisions returns the set of outstanding revisions that have not been
+// applied, after optionally defining a new revision of outstanding mutations.
 func (s *Server) DefineRevisions(ctx context.Context,
 	in *spb.DefineRevisionsRequest) (*spb.DefineRevisionsResponse, error) {
-	// Get the previous and current high water marks.
-	mapClient, err := s.trillian.MapClient(ctx, in.DirectoryId)
-	if err != nil {
-		return nil, err
-	}
-	_, latestMapRoot, err := mapClient.GetAndVerifyLatestMapRoot(ctx)
-	if err != nil {
-		return nil, err
-	}
-
-	// Collect a list of unapplied revisions.
-	highestRev, err := s.batcher.HighestRev(ctx, in.DirectoryId)
+	// Get the list of unprocessed revisions.
+	root, high, err := s.getRootAndRev(ctx, in.DirectoryId)
 	if err != nil {
 		return nil, err
 	}
-	outstanding := []int64{}
-	for rev := int64(latestMapRoot.Revision) + 1; rev <= highestRev; rev++ {
-		outstanding = append(outstanding, rev)
+	resp := &spb.DefineRevisionsResponse{
+		Begin: int64(root.Revision) + 1,
+		End:   high,
 	}
 
 	// Don't create new revisions if there are ones waiting to be applied.
-	if len(outstanding) > 0 {
-		return &spb.DefineRevisionsResponse{OutstandingRevisions: outstanding}, nil
+	// TODO(#1056): Allow going ahead to some extent.
+	if resp.End > resp.Begin {
```

Comment:
We probably should allow at lest 1 outstanding revision here, otherwise this PR will only make things 1.5 times slower.